interfaces/apparmor/backend.go

```diff
@@ -0,0 +1,175 @@
+//
+// NOTE: A systemd job (TODO: specify which) loads all snappy-specific apparmor
+// profiles into the kernel during the boot process.
+package apparmor
+
+import (
+	"bytes"
+	"fmt"
+	"path/filepath"
+
+	"github.com/ubuntu-core/snappy/dirs"
+	"github.com/ubuntu-core/snappy/interfaces"
+	"github.com/ubuntu-core/snappy/osutil"
+	"github.com/ubuntu-core/snappy/snap"
+)
+
+// Backend is responsible for maintaining apparmor profiles for ubuntu-core-launcher.
+type Backend struct {
+	// Custom template exists to support old-security which goes
+	// beyond what is possible with pure security snippets.
+	//
+	// If non-empty then it overrides the built-in template.
+	CustomTemplate string
+}
+
+// Configure creates and loads security artefacts specific to a given snap. The
+// snap can be in developer mode to make security violations non-fatal to the
+// offending application process.
+//
+// This method should be called after changing plug, slots, connections between
+// them or application present in the snap.
+func (b *Backend) Configure(snapInfo *snap.Info, repo *interfaces.Repository, developerMode bool) error {
+	// Get the snippets that apply to this snap
+	snippets, err := repo.SecuritySnippetsForSnap(snapInfo.Name, b.SecuritySystem())
+	if err != nil {
+		return fmt.Errorf("cannot obtain security snippets for snap %q: %s", snapInfo.Name, err)
+	}
+	// Get the files that this snap should have
+	content, err := b.CombineSnippets(snapInfo, developerMode, snippets)
+	if err != nil {
+		return fmt.Errorf("cannot obtain expected security files for snap %q: %s", snapInfo.Name, err)
+	}
+	changed, removed, err := osutil.EnsureDirState(b.Directory(), b.FileGlob(snapInfo), content)
+	if err != nil {
+		return fmt.Errorf("cannot synchronize security files for snap %q: %s", snapInfo.Name, err)
+	}
+	err = b.ObserveChanges(changed, removed)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Deconfigure removes security artefacts of a given snap.
+func (b *Backend) Deconfigure(snapInfo *snap.Info) error {
+	_, removed, err := osutil.EnsureDirState(b.Directory(), b.FileGlob(snapInfo), nil)
+	if err != nil {
+		return fmt.Errorf("cannot synchronize security files for snap %q: %s", snapInfo.Name, err)
+	}
+	err = b.ObserveChanges(nil, removed)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// CombineSnippets combines security snippets collected from all the interfaces
+// affecting a given snap into a content map applicable to EnsureDirState. The
+// backend delegates writing those files to higher layers.
+func (b *Backend) CombineSnippets(snapInfo *snap.Info, developerMode bool, snippets map[string][][]byte) (content map[string]*osutil.FileState, err error) {
+	for _, appInfo := range snapInfo.Apps {
+		s := make([][]byte, 0, len(snippets[appInfo.Name])+2)
+		s = append(s, b.aaHeader(appInfo, developerMode))
+		s = append(s, snippets[appInfo.Name]...)
+		s = append(s, []byte("}\n"))
+		fileContent := bytes.Join(s, []byte("\n"))
```

Comment:
The logic here is excessively wasteful. Allocate a slice of slices to append slices to reallocate a slice to then join the slices. Let's please do it all in one shot: allocate once with the final length, write the final content into it.